core/src/main/java/nu/yona/server/analysis/entities/WeekActivity.java

```diff
@@ -104,22 +104,33 @@ public static WeekActivity createInstance(UserAnonymized userAnonymized, Goal go
 	public static WeekActivity createInstanceInactivity(UserAnonymized userAnonymized, Goal goal, ZonedDateTime startOfWeek)
 	{
 		WeekActivity result = createInstance(userAnonymized, goal, startOfWeek);
-		// add the days
-		// notice this doesn't take care of user time zone changes during the week
-		// so for consistency it is important that the batch script adding inactivity does so
-		for (int i = 0; i < 7; i++)
+		result.addInactivityDays();
+		return result;
+	}
+
+	public void addInactivityDays() 
+	{
+		// if the batch job has already run, skip
+		if(dayActivities.size() < 7)
 		{
-			ZonedDateTime startOfDay = startOfWeek.plusDays(i);
-			if (isInFuture(startOfDay, startOfWeek.getZone()))
+			// notice this doesn't take care of user time zone changes during the week
+			// so for consistency it is important that the batch script adding inactivity does so
+			for (int i = 0; i < 7; i++)
 			{
-				break;
-			}
-			if (goal.wasActiveAtInterval(startOfDay, ChronoUnit.DAYS))
-			{
-				result.addDayActivity(DayActivity.createInstanceInactivity(userAnonymized, goal, startOfDay));
+				ZonedDateTime startOfDay = getStartTime().plusDays(i);
+				if (isInFuture(startOfDay, getStartTime().getZone()))
+				{
+					break;
+				}
+				if (getGoal().wasActiveAtInterval(startOfDay, ChronoUnit.DAYS))
+				{
+					if(!dayActivities.stream().anyMatch(dayActivity -> dayActivity.getDate().getDayOfWeek().equals(startOfDay.getDayOfWeek())))
```

Comment:
You're now here testing whether the day is already available in the list, right? You're maintaining the day activities in a list, so wouldn't it be a lot simpler of position 0 of the list contains Sunday, position 1 contains Monday, etc.? The you could put a marker on ``WeekActivity`` to indicate the week is completely filled, as an alternative to the ``if(dayActivities.size() < 7)`` test, and for the rest simply use the week day as index in the list.